File: frontend/src/types/investigations.ts
// src/types/investigations.ts
// Wire shapes mirror the workbench backend exactly (services/workbench/schemas/investigations.py,
// comments.py, timeline.py).

export type InvestigationStatus =
  | 'open'
  | 'active'
  | 'awaiting_information'
  | 'submitted'
  | 'returned'
  | 'completed'
  | 'cancelled';

export type InvestigationPriority = 'critical' | 'high' | 'medium' | 'low';

/** findings_refs entry — structured reference [{ type, id, description }]. */
export interface FindingRef {
  type?: string;
  id?: string;
  description?: string;
}

export interface Investigation {
  investigation_id: string;
  title: string;
  description?: string | null;
  alert_id?: string | null;
  scope_id: string;
  status: InvestigationStatus;
  priority: InvestigationPriority;
  assigned_to?: string | null;
  created_by: string;
  findings_text?: string | null;
  findings_refs?: FindingRef[] | null;
  conclusion?: string | null;
  started_at?: string | null;
  submitted_at?: string | null;
  completed_at?: string | null;
  return_reason?: string | null;
  version: number;
  created_at: string;
  updated_at: string;
}

export interface InvestigationListResponse {
  total: number;
  page: number;
  page_size: number;
  items: Investigation[];
}

export interface InvestigationMutationResponse {
  success: boolean;
  investigation: Investigation;
  version: number;
}

export interface Comment {
  comment_id: string;
  entity_type: string;
  entity_id: string;
  content?: string;
  author_id: string;
  is_internal: boolean;
  is_redacted: boolean;
  redacted_at?: string | null;
  redacted_by?: string | null;
  version: number;
  created_at: string;
  updated_at: string;
}

export interface CommentListResponse {
  total: number;
  page: number;
  page_size: number;
  items: Comment[];
}

export interface CommentMutationResponse {
  success: boolean;
  comment: Comment;
  version: number;
}

export interface TimelineEntry {
  timeline_id: string;
  entity_type: string;
  entity_id: string;
  event_type: string;
  actor_id: string;
  old_value?: Record<string, unknown> | null;
  new_value?: Record<string, unknown> | null;
  metadata?: Record<string, unknown> | null;
  occurred_at: string;
}

export interface TimelineListResponse {
  total: number;
  page: number;
  page_size: number;
  items: TimelineEntry[];
}

File: services/workbench/schemas/investigations.py
from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any, List, Optional

from pydantic import BaseModel, Field


class InvestigationStatus(str, Enum):
    OPEN = "open"
    ACTIVE = "active"
    AWAITING_INFORMATION = "awaiting_information"
    SUBMITTED = "submitted"
    RETURNED = "returned"
    COMPLETED = "completed"
    CANCELLED = "cancelled"


class InvestigationPriority(str, Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class UpdateInvestigationRequest(BaseModel):
    findings_text: Optional[str] = None
    findings_refs: Optional[List[Any]] = None
    conclusion: Optional[str] = None
    expected_version: int = Field(ge=1)


class TransitionInvestigationRequest(BaseModel):
    target_status: str
    return_reason: Optional[str] = None
    expected_version: int = Field(ge=1)


class CancelInvestigationRequest(BaseModel):
    cancel_reason: str = Field(min_length=1)
    expected_version: int = Field(ge=1)


class ReviewNotHarmfulRequest(BaseModel):
    rationale: str = Field(min_length=1)
    expected_version: int = Field(ge=1)


class EscalateInvestigationRequest(BaseModel):
    title: str = Field(min_length=1)
    priority: str = "medium"
    rationale: str = Field(min_length=1)
    expected_version: int = Field(ge=1)


class InvestigationResponse(BaseModel):
    investigation_id: str
    title: str
    description: Optional[str] = None
    alert_id: Optional[str] = None
    scope_id: str
    status: str
    priority: str
    assigned_to: Optional[str] = None
    created_by: str
    findings_text: Optional[str] = None
    findings_refs: Optional[List[Any]] = None
    conclusion: Optional[str] = None
    started_at: Optional[datetime] = None
    submitted_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    return_reason: Optional[str] = None
    version: int
    created_at: datetime
    updated_at: datetime


class InvestigationListResponse(BaseModel):
    total: int
    page: int
    page_size: int
    items: List[InvestigationResponse]


class InvestigationMutationResponse(BaseModel):
    success: bool = True
    investigation: InvestigationResponse
    version: int


class EscalateInvestigationResponse(BaseModel):
    success: bool = True
    investigation: InvestigationResponse
    case_id: str
    version: int

